include/ignition/math/detail/InterpolationPoint.hh

```diff
@@ -0,0 +1,251 @@
+    {
+      /// \brief The position of the point
+      Vector3<T> position;
+
+      /// \brief The index from which this point was retrieved.
+      /// Can be used by the application calling it to index. The reason
+      /// this is optional is that data may be missing from a sparse grid.
+      std::optional<std::size_t> index;
+    };
+
+    /// \brief Describes an interpolation point in1d.
+    template<typename T>
+    struct InterpolationPoint1D
+    {
+      /// \brief The position of the point
+      T position;
+
+      /// \brief The index from which this point was retrieved.
+      /// Can be used by the application calling it to index.
+      std::size_t index;
+    };
+
+    /// \brief Linear Interpolation of two points along a 1D line.
+    /// \param[in] _a The first point.
+    /// \param[in] _b The second point.
+    /// \param[in] _lst An array of values that are to be used by the
+    /// interpolator. _lst[a.index] and _lst[b.index] are the values
+    /// to be interpolated.
+    /// \param[in] _pos The position to interpolate.
+    /// Warning: This function assumes that the indices of _a and _b correspond
+    /// to values in _lst. It performs no bounds checking whatsoever and if you
+    /// pass it invalid data, it will crash. It also assumes that _a and _b
+    /// are not in the same position.
+    template<typename T, typename V>
+    V LinearInterpolate(
+      const InterpolationPoint1D<T> &_a,
+      const InterpolationPoint1D<T> &_b,
+      const std::vector<V>  &_lst,
+      const T &_pos
+      )
+    {
+      assert(abs(_a.position - _b.position) > 0);
+      assert(_a.index < _lst.size());
+      assert(_b.index < _lst.size());
+
+      auto t =  (_pos - _b.position)/ (_a.position - _b.position);
+      return (1 - t) * _lst[_b.index] + t * _lst[_a.index];
+    }
+
+    /// \brief Linear Interpolation of two points in 3D space
+    /// \param[in] _a The first point.
+    /// \param[in] _b The second point.
+    /// \param[in] _lst An array of values that are to be used by the
+    /// interpolator. _lst[a.index] and _lst[b.index] are the values
+    /// to be interpolated. If a.index or b.index is std::nullopt then use the
+    /// default value.
+    /// \param[in] _pos The position to interpolate.
+    /// \param[in] _default The default value to use if a.index or b.index is
+    /// std::nullopt.
+    /// Warning: This function assumes that the indices of _a and _b correspond
+    /// to values in _lst. It performs no bounds checking whatsoever and if you
+    /// pass it invalid data, it will crash.
+    template<typename T, typename V>
+    V LinearInterpolate(
+      const InterpolationPoint3D<T> &_a,
+      const InterpolationPoint3D<T> &_b,
+      const std::vector<V>  &_lst,
+      const Vector3<T> &_pos,
+      const V &_default = V(0)
+      )
+    {
+      assert(abs((_a.position - _b.position).Length()) > 0);
+      assert((_a.index.has_value()) ? _a.index.value() < _lst.size(): true);
```

Comment:
@arjo129 nit^N!: 
```suggestion
      assert(_a.index.value_or(-1) < _lst.size());
```
would be equivalent.

Reply:
I think this abuses sign vs unsigned int comparisons.

Reply:
Fair enough.